pkg/reconciler/ingress/ingress.go

```diff
@@ -148,83 +148,84 @@ func (c *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
+// LoadBalancerIngressStatuses for the current Gateways in use.
+func (c *Reconciler) lookUpLoadBalancers(ing *v1alpha1.Ingress, gpc *config.GatewayPlugin) (map[v1alpha1.IngressVisibility][]v1alpha1.LoadBalancerIngressStatus, error) {
+	externalStatuses, err := c.collectLBIngressStatus(ing, gpc.ExternalGateways)
+	if err != nil {
+		return nil, err
 	}
-	gw, err := c.gatewayLister.Gateways(gwc.Namespace).Get(gwc.Name)
+
+	internalStatuses, err := c.collectLBIngressStatus(ing, gpc.LocalGateways)
 	if err != nil {
 		return nil, err
 	}
 
-	var lbis v1alpha1.LoadBalancerIngressStatus
+	return map[v1alpha1.IngressVisibility][]v1alpha1.LoadBalancerIngressStatus{
+		v1alpha1.IngressVisibilityExternalIP:   externalStatuses,
+		v1alpha1.IngressVisibilityClusterLocal: internalStatuses,
+	}, nil
+}
 
-	if len(gw.Status.Addresses) > 0 {
-		switch *gw.Status.Addresses[0].Type {
-		case gatewayapi.IPAddressType:
-			lbis = v1alpha1.LoadBalancerIngressStatus{IP: gw.Status.Addresses[0].Value}
-		default:
-			// Should this actually be under Domain? It seems like the rest of the code expects DomainInternal though...
-			lbis = v1alpha1.LoadBalancerIngressStatus{DomainInternal: gw.Status.Addresses[0].Value}
+// collectLBIngressStatus will return LoadBalancerIngressStatuses for the
+// provided single Gateway config. If a service is available on a Gateway, it will
+// return the address of the service.  Otherwise, it will return the first
+// address in the Gateway status.
+func (c *Reconciler) collectLBIngressStatus(ing *v1alpha1.Ingress, gatewayConfigs []config.Gateway) ([]v1alpha1.LoadBalancerIngressStatus, error) {
+	statuses := []v1alpha1.LoadBalancerIngressStatus{}
+
+	// TODO: currently only 1 gateway is supported. When the config is updated to
+	// support multiple, this code must change to find out which Gateway is
+	// appropriate for the given Ingress
+	gwc := gatewayConfigs[0]
+	if gwc.Service != nil {
+		statuses = append(statuses, v1alpha1.LoadBalancerIngressStatus{
+			DomainInternal: network.GetServiceHostname(gwc.Service.Name, gwc.Service.Namespace),
+		})
+	} else {
+		gw, err := c.gatewayLister.Gateways(gwc.Namespace).Get(gwc.Name)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				ing.Status.MarkLoadBalancerFailed(
+					"GatewayDoesNotExist",
+					fmt.Sprintf(
+						"could not find Gateway %s/%s",
+						gwc.Namespace,
+						gwc.Name,
+					),
+				)
+			}
+			return nil, err
 		}
 
-		return []v1alpha1.LoadBalancerIngressStatus{lbis}, nil
+		if len(gw.Status.Addresses) > 0 {
+			switch *gw.Status.Addresses[0].Type {
+			case gatewayapi.IPAddressType:
+				statuses = append(statuses, v1alpha1.LoadBalancerIngressStatus{IP: gw.Status.Addresses[0].Value})
+			default:
+				// Should this actually be under Domain? It seems like the rest of the code expects DomainInternal though...
+				statuses = append(statuses, v1alpha1.LoadBalancerIngressStatus{DomainInternal: gw.Status.Addresses[0].Value})
```

Comment:
I think this should actually be `Domain` - but honestly I don't know how serving handles that - so I'd say let's leave this for now and adjust in a follow up once we see what serving does

Reply:
yeah I'm not sure. I had it as Domain originally, but when @andrew-su and @nader-ziada were testing things, they found it didn't behave correctly. So I stuck with DomainInternal

Reply:
yeah - so there's followup in serving that'll have to happen - leave it as DomainInternal for now